src/server/coder.ts

```diff
@@ -0,0 +1,360 @@
+    if (typeof responseJson === 'object' && responseJson !== null && 'data' in responseJson) {
+        organizations = responseJson.data
+    }
+    // If it's an object with organizations property
+    else if (typeof responseJson === 'object' && responseJson !== null && 'organizations' in responseJson) {
+        organizations = responseJson.organizations
+    }
+
+    // Ensure we have an array
+    if (!Array.isArray(organizations)) {
+        // If it's still not an array, try to extract first array value from object
+        if (typeof responseJson === 'object' && responseJson !== null) {
+            const arrayValues = Object.values(responseJson).filter(Array.isArray)
+            if (arrayValues.length > 0) {
+                organizations = arrayValues[0]
+            } else {
+                // Last resort: convert to array if it's a single object
+                if (typeof responseJson === 'object' && responseJson !== null) {
+                    organizations = [responseJson]
+                } else {
+                    throw new Error('Failed to extract organizations array from response')
+                }
+            }
+        } else {
+            throw new Error('Failed to extract organizations array from response')
+        }
+    }
+
+    const foundOrg = organizations.find((org: CoderBaseEntity) => org.name === 'coder')
+    if (!foundOrg) {
+        throw new Error('Coder organization not found')
+    }
+    return foundOrg.id
+}
+
+export const getCoderTemplateId = async () => {
+    const coderApiEndpoint = await getConfigValue('CODER_API_ENDPOINT')
+    const coderToken = await getConfigValue('CODER_TOKEN')
+    const coderTemplate = await getConfigValue('CODER_TEMPLATE')
+    const templatesResponse = await fetch(`${coderApiEndpoint}${coderTemplateId()}`, {
+        method: 'GET',
+        headers: {
+            Accept: 'application/json',
+            'Coder-Session-Token': coderToken,
+        },
+    })
+    if (!templatesResponse.ok) {
+        throw new Error('Failed to fetch templates data from Coder API')
+    }
+    const responseJson = await templatesResponse.json()
+    // Handle different response structures - sometimes it's an object with data field, sometimes direct array
+    const templates = Array.isArray(responseJson) ? responseJson : responseJson.data || responseJson
+    const foundTemplate = templates.find((template: CoderBaseEntity) => template.name === coderTemplate)
+    if (!foundTemplate) {
+        throw new Error(`Template with name '${coderTemplate}' not found`)
+    }
+    return foundTemplate.id
+}
+
+export function shaHash(input: string): string {
+    return createHash('sha256').update(input).digest('hex')
+}
+
+export function generateWorkspaceName(studyId: string) {
+    if (!studyId) return ''
+    const result = shaHash(studyId.replaceAll(/[^a-z0-9]/gi, '').toLowerCase())
+    return result.substring(0, CODER_WORKSPACE_NAME_LIMIT)
+}
```

Comment:
I stand by my earlier comment that this is _way_ too complex and would really benefit from a REST api wrapper 😁 

But I think we can let this go and attempt to simplify it later, at least it's contained to this file